virtualizarr/storage/_obstore.py

```diff
@@ -0,0 +1,246 @@
+    get_zarr_metadata,
+    list_dir_from_xr_obj,
+    parse_manifest_index,
+)
+
+if TYPE_CHECKING:
+    from collections.abc import AsyncGenerator, Iterable
+    from typing import Any
+
+    from obstore.store import ObjectStore as _UpstreamObjectStore
+    from zarr.core.buffer import BufferPrototype
+    from zarr.core.common import BytesLike
+
+
+__all__ = ["VirtualObjectStore"]
+
+_ALLOWED_EXCEPTIONS: tuple[type[Exception], ...] = (
+    FileNotFoundError,
+    IsADirectoryError,
+    NotADirectoryError,
+)
+
+
+class VirtualObjectStore(Store):
+    """A Zarr store that uses obstore for fast read/write from AWS, GCP, Azure.
+
+    Parameters
+    ----------
+    stores : dict[prefix, obstore.store.ObjectStore]
+        A mapping of url prefixes to obstore Store instances set up with the proper credentials.
+
+    Warnings
+    --------
+    VirtualObjectStore is experimental and subject to API changes without notice. Please
+    raise an issue with any comments/concerns about the store.
+
+    Notes
+    -----
+    Modified from https://github.com/zarr-developers/zarr-python/pull/1661
+    """
+
+    def __eq__(self, value: object):
+        NotImplementedError
+
+    def __init__(
+        self,
+        xr_obj: DataArray | Dataset,
+        stores: dict[str, _UpstreamObjectStore],
+    ) -> None:
+        import obstore as obs
+
+        # TODO: Support DataArray, Dataset, or DataTree across all methods
+        for store in stores.values():
+            if not isinstance(
+                store,
+                (
+                    obs.store.AzureStore,
+                    obs.store.GCSStore,
+                    obs.store.HTTPStore,
+                    obs.store.S3Store,
+                    obs.store.LocalStore,
+                    obs.store.MemoryStore,
+                ),
+            ):
+                raise TypeError(f"expected ObjectStore class, got {store!r}")
+        super().__init__(read_only=True)
+        self.stores = stores
+        self.xr_obj: DataArray | Dataset = xr_obj
+
+    def __str__(self) -> str:
+        return f"ManifesStore({self.xr_obj})"
+
+    def __getstate__(self) -> dict[Any, Any]:
+        state = self.__dict__.copy()
+        stores = state["stores"]
+        for k, v in stores:
+            stores[k] = pickle.dumps(v)
+        state["stores"] = stores
+        return state
+
+    def __setstate__(self, state: dict[Any, Any]) -> None:
+        stores = state["stores"]
+        for k, v in stores:
+            stores[k] = pickle.loads(v)
+        state["stores"] = stores
+        self.__dict__.update(state)
+
+    async def get(
+        self,
+        key: str,
+        prototype: BufferPrototype,
+        byte_range: ByteRequest | None = None,
+    ) -> Buffer | None:
+        # docstring inherited
+        import obstore as obs
+
+        if "zarr.json" in key:
+            return get_zarr_metadata(self.xr_obj, key)
+        manifest_index = parse_manifest_index(key)
```

Comment:
I don't think I really understand the need for this `ManifestIndex` class. All it saves you is returning two separate variables in this line (`var_name` & `chunk_key`).